File: src/jray/common/Sky.java
package jray.common;

import java.io.IOException;

import jray.math.Vect;

public class Sky extends Object3D {

    Texture texture;

    public Sky(String texture) throws IOException {
        super(null, null);
        this.texture = Texture.loadTexture(texture);
    }

    public Sky(int color) {
        super(null, null, 0);
        this.color = color;
    }

    @Override
    public double getHitPointDistance(Ray r) {
        return Double.POSITIVE_INFINITY;
    }

    @Override
    public int getColorAt(Vect3 hitPoint) {
        double[] hpdat = hitPoint.getData();

        double x = Math.acos(hpdat[1] / hitPoint.length()) / Math.PI;
        double y = Math.acos(hpdat[2] / hitPoint.length()) / Math.PI;

        return texture.getColorAt(x, y);
    }

    @Override
    public void getNormalAt(Vect3 hitPoint, Vect3 normal) {
        Vect.invert(hitPoint, normal); // the normal is everywhere the vect pt -> 0
    }

    @Override
    public boolean contains(Vect3 hitPoint) {
        return false;
    }

    @Override
    public void rotate(Matrix4 rotationMatrix) {
    }
}

File: JRayXLib/JRayXLib/Common/Sky.cs
using JRayXLib.Math;

namespace JRayXLib.Common
{
    public class Sky : Object3D {
        readonly Texture _texture;

        public Sky(string texture) : base(null, null){
            _texture = Texture.Load(texture);
        }

        public Sky(uint color) : base(null, null, 0){
            Color = color;
        }

        public override double GetHitPointDistance(Ray r) {
            return double.PositiveInfinity;
        }

        public new int GetColorAt(Vect3 hitPoint) {
            double[] hpdat = hitPoint.GetData();

            double x = System.Math.Acos(hpdat[1] / hitPoint.Length()) / System.Math.PI;
            double y = System.Math.Acos(hpdat[2] / hitPoint.Length()) / System.Math.PI;

            return _texture.GetColorAt(x, y);
        }

        public override void GetNormalAt(Vect3 hitPoint, Vect3 normal) {
            Vect.Invert(hitPoint, normal); // the normal is everywhere the vect pt -> 0
        }

        public override bool Contains(Vect3 hitPoint) {
            return false;
        }

        public override void Rotate(Matrix4 rotationMatrix) {
        }
    }
}
